### ISSUE TITLE
QUnit to mocha and/or jest?

### ISSUE THREAD

**Issue by peripateticus** [luijar/functional-programming-js#12]
Hi Luis,

Just wondering what your thoughts were on getting onto mocha/istanbul test stack rather than QUnit/Blanket?  I've forked and created a branch [here](https://github.com/peripateticus/functional-programming-js/blob/qunit-to-mocha/package.json#L27) to illustrate with Chapters 3 and 4.

Some advantages to this are:

- Running *all* unit tests in one command as well as individual chapters. 
- Test coverage across all unit tests.
- Detailed test coverage reports.
![screen shot 2017-07-31 at 2 46 26 pm](https://user-images.githubusercontent.com/1110797/28799908-47014082-75ff-11e7-8f48-90f182fb69ab.png)
![screen shot 2017-07-31 at 2 50 42 pm](https://user-images.githubusercontent.com/1110797/28800024-c3354b76-75ff-11e7-9b47-e6bcf483fc15.png)

- `watch` mechanism built in for instant feedback while working with code.
- `mocha`/`istanbul` are ubiquitous now for unit testing in `nodejs`.

Here's sample output:

```
$ npm test                                                            [ 2:29PM Mon Jul 31 ]

fp-js-manning@2.0.0 test /Users/aperkin/prj/functional-programming-js
mocha src/**/ch*-tests.js

  Chapter 3
    ✓ Understanding reduce
    ✓ Combining map and reduce
    ✓ Combining map and reduce with lenses
    ✓ Valid or not valid
    ✓ Introducing filter
    ✓ People born in 1903
    ✓ Array processing with Lodash
    ✓ Gather stats
    ✓ Lazy function chains
    ✓ SQL-like JavaScript
    ✓ Recursive addition
    ✓ Recursive addition 2
    ✓ Tree navigation

  Chapter 4
    ✓ Chaining methods together
    ✓ Check Type tests
    ✓ Tuple test
    ✓ Extending the core language
    ✓ Compsition
    ✓ More compsition
    ✓ Composition with functional libraries
    ✓ Composition as point-free functions
    ✓ Show student program with currying and composition
    ✓ More point-free coding


  23 passing (24ms)
```

```
$ npm run ch04                                                   [ 2:29PM Mon Jul 31 ]

fp-js-manning@2.0.0 ch04 /Users/aperkin/prj/functional-programming-js
mocha src/ch04/ch04-tests.js

  Chapter 4
    ✓ Chaining methods together
    ✓ Check Type tests
    ✓ Tuple test
    ✓ Extending the core language
    ✓ Compsition
    ✓ More compsition
    ✓ Composition with functional libraries
    ✓ Composition as point-free functions
    ✓ Show student program with currying and composition
    ✓ More point-free coding


  10 passing (16ms)
```

```
$ npm run cover
fp-js-manning@2.0.0 cover /Users/aperkin/prj/functional-programming-js
istanbul cover node_modules/.bin/_mocha src/**/ch*-tests.js

  Chapter 3
    ✓ Understanding reduce
    ✓ Combining map and reduce
    ✓ Combining map and reduce with lenses
    ✓ Valid or not valid
    ✓ Introducing filter
    ✓ People born in 1903
    ✓ Array processing with Lodash
    ✓ Gather stats
    ✓ Lazy function chains
    ✓ SQL-like JavaScript
    ✓ Recursive addition
    ✓ Recursive addition 2
    ✓ Tree navigation

  Chapter 4
    ✓ Chaining methods together
    ✓ Check Type tests
    ✓ Tuple test
    ✓ Extending the core language
    ✓ Compsition
    ✓ More compsition
    ✓ Composition with functional libraries
    ✓ Composition as point-free functions
    ✓ Show student program with currying and composition
    ✓ More point-free coding


  23 passing (42ms)

=============================================================================
Writing coverage object [/Users/aperkin/prj/functional-programming-js/coverage/coverage.json]
Writing coverage reports at [/Users/aperkin/prj/functional-programming-js/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 95.82% ( 321/335 )
Branches     : 88.46% ( 46/52 )
Functions    : 87.69% ( 57/65 )
Lines        : 96.08% ( 294/306 )
================================================================================

```

Just some thoughts...

**Comment by peterschussheim** [comment]
Any reason you aren't using Jest?  I've found It to be faster and easier to use and the developer experience (in my opinion) is better compared to mocha.

**Comment by peripateticus** [comment]
`Jest` would be great as well, either/or. Jest has the additional React functionality that isn't really  needed but in terms of achieving the same functionality, it would totally work and bonus of built-in coverage.

**Comment by JasonSooter** [comment]
I've been converting the tests to Jest along the way. It's a fairly easy switch out but I'd vote for Jest as well!

**Comment by peripateticus** [comment]
Create a `jest` branch [here](https://github.com/peripateticus/functional-programming-js/tree/qunit-to-jest) and refactored [ch03](https://github.com/peripateticus/functional-programming-js/blob/qunit-to-jest/src/ch03/ch03.test.js) tests. So the output looks very similar to mocha output:

```
$ npm test                                                            [10:09AM Tue Aug  1 ]

> fp-js-manning@2.0.0 test /Users/aperkin/prj/functional-programming-js
> jest

 PASS  src/ch03/ch03.test.js
  Chapter 3
    ✓ Understanding reduce (3ms)
    ✓ Combining map and reduce
    ✓ Combining map and reduce with lenses (1ms)
    ✓ Valid or not valid (1ms)
    ✓ Introducing filter (1ms)
    ✓ People born in 1903
    ✓ Array processing with Lodash (1ms)
    ✓ Gather stats (1ms)
    ✓ Lazy function chains (2ms)
    ✓ SQL-like JavaScript
    ✓ Recursive addition (1ms)
    ✓ Recursive addition 2
    ✓ Tree navigation (2ms)

Test Suites: 1 passed, 1 total
Tests:       13 passed, 13 total
Snapshots:   0 total
Time:        0.672s, estimated 1s
Ran all test suites.
```

```
$ npm run cover                                                       [10:09AM Tue Aug  1 ]

> fp-js-manning@2.0.0 cover /Users/aperkin/prj/functional-programming-js
> jest --coverage --coverageDirectory=coverage

 PASS  src/ch03/ch03.test.js
  Chapter 3
    ✓ Understanding reduce (5ms)
    ✓ Combining map and reduce (1ms)
    ✓ Combining map and reduce with lenses (2ms)
    ✓ Valid or not valid (1ms)
    ✓ Introducing filter
    ✓ People born in 1903 (1ms)
    ✓ Array processing with Lodash (1ms)
    ✓ Gather stats (1ms)
    ✓ Lazy function chains (2ms)
    ✓ SQL-like JavaScript
    ✓ Recursive addition
    ✓ Recursive addition 2
    ✓ Tree navigation (5ms)

Test Suites: 1 passed, 1 total
Tests:       13 passed, 13 total
Snapshots:   0 total
Time:        0.7s, estimated 1s
Ran all test suites.
-------------|----------|----------|----------|----------|----------------|
File         |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-------------|----------|----------|----------|----------|----------------|
All files    |    79.03 |    89.47 |    63.33 |    80.33 |                |
 ch03/model  |    88.57 |    84.62 |    73.33 |    91.18 |                |
  Node.js    |    81.25 |        0 |    66.67 |    86.67 |          17,43 |
  Tree.js    |    94.74 |      100 |    83.33 |    94.74 |             27 |
 model       |    66.67 |      100 |    53.33 |    66.67 |                |
  Address.js |    61.54 |      100 |    42.86 |    61.54 | 25,33,37,45,46 |
  Person.js  |    71.43 |      100 |     62.5 |    71.43 |    16,24,25,29 |
-------------|----------|----------|----------|----------|----------------|
```
`jest` tests do run slower than mocha tests (and I'm running both ch03 and ch04 tests in `mocha` but only ch03 tests in `jest` yet `jest` is still quite a bit slower. You can run both branches with a `time npm test`.) But `jest` is more feature rich.  Pros and cons. I did run into [this issue](https://github.com/facebook/jest/issues/1767) when running `jest --watch` but the resolution in the thread resolved it.

**Comment by peripateticus** [comment]
I think the bigger concern that I had is the fact that the book makes several references `QUnit` so the test code now differs from the book code. I don't know what all of the ramifications are of doing something like this. Does it matter or can a note be included in the repo to inform the user of the change? Or does the book need to be updated at some future edition that reflects the code? It's really not a huge change in that it's abstracted away when running the npm scripts and the underlying syntax of these test runners are *really* similar

**Comment by luijar** [comment]
This looks great @peripateticus (and thanks everyone else for chiming in). Yeah, so what prevents me from doing it is what you said exactly, its kind of written in stone to use QUnit. Again, the purpose was more on the FP concepts, and much less on what framework to use. In chap 7, I picked QUnit cause I atm thought it was easier and plugged into to blanket easily. Perhaps we could just create a copy of that chapter (and any other) folder and use something else, then make a note in the README, I think that'd be fine. If we add to it, it's probably OK instead of replacing. In hindsight, I did end up switching to Mocha/Chai for unit testing, been meaning to look at Jest (looks great). 
So, we could try to just create another folder (per chapter) to use a different testing framework, that's fine. Readers will see a flavor of using both, which couldn't hurt :-)
